renku/core/models/dataset.py

```diff
@@ -538,70 +509,82 @@ def update_files_from(self, current_files: List[DatasetFile], date: datetime = N
+        """Updates metadata."""
+        editable_attributes = ["creators", "description", "keywords", "title"]
+        for name, value in kwargs.items():
+            if name not in editable_attributes:
+                raise errors.ParameterError(f"Cannot edit field: '{name}'")
+            if value and value != getattr(self, name):
+                setattr(self, name, value)
+
+    def unlink_file(self, path, missing_ok=False) -> Optional[DatasetFile]:
+        """Mark a file as removed using its relative path."""
+        assert not self.immutable, f"Dataset is immutable {self}"
+
+        file = self.find_file(path)
+
+        if not file:
+            if not missing_ok:
+                raise errors.InvalidFileOperation(f"File cannot be found: {path}")
+            return
 
-        return old_datasets.Dataset(
-            name=self.name,
-            client=client,
-            creators=convert_creators(self.creators),
-            date_created=self.date_created,
-            date_published=self.date_published,
-            derived_from=derived_from,
-            description=self.description,
-            files=self._convert_to_dataset_files(client),
-            id=None,
-            identifier=self.identifier,
-            images=[image.to_image_object() for image in self.images] if self.images else None,
-            in_language=self.in_language.to_language() if self.in_language else None,
-            keywords=self.keywords,
-            license=self.license,
-            project=client._project,
-            same_as=self.same_as.to_url(client) if self.same_as else None,
-            tags=[tag.to_dataset_tag(client) for tag in self.tags],
-            title=self.title,
-            url=None,
-            version=self.version,
-        )
+        file.remove()
+
+        return file
 
-    def _convert_to_dataset_files(self, client):
-        """Create instances of renku.core.models.datasets.DatasetFile."""
-        dataset_files = []
-        for file in self.files:
-            dataset_file = file.to_dataset_file(client)
-            if not dataset_file:
-                continue
+    def add_or_update_files(self, files: Union[DatasetFile, List[DatasetFile]]):
+        """Add new files or update existing files."""
+        assert not self.immutable, f"Dataset is immutable {self}"
 
-            dataset_files.append(dataset_file)
+        if isinstance(files, DatasetFile):
+            files = [files]
 
-        return dataset_files
+        new_files = []
+
+        for file in files:
+            existing_file = self.find_file(file.entity.path)
+            if not existing_file:
+                new_files.append(file)
+            elif file.entity.checksum != existing_file.entity.checksum or file.date_added != existing_file.date_added:
+                self.dataset_files.remove(existing_file)
+                new_files.append(file)
+
+        if not new_files:
+            return
+
+        self.dataset_files += new_files
```

Comment:
It will generate a separate Persistent object in the database. Synching these 2 objects and making sure that the list is immutable is more difficult than just making sure that we set `_p_changed` when files changes.